### PR TITLE
Ability to use saved display filters

### PR DESCRIPTION
**Pull request by vapniks**

### CODE REVIEW

**Comment by vapniks** [comment]
Added command for reading ~/wireshark/dfilters file and prompting for a named display filter.
Also quick convenience function for clearing the current filter.

File: pcap-mode.el [hunk 1]
```diff
@@ -112,14 +122,16 @@
...
     (define-key kmap (kbd "f") 'pcap-mode-set-tshark-filter)
+    (define-key kmap (kbd "F") 'pcap-mode-set-named-filter)
     (define-key kmap (kbd "c") 'pcap-mode-search-frames)
     (define-key kmap (kbd "\C-u f")
       'pcap-mode-set-tshark-single-packet-filter)
     (define-key kmap (kbd "s") 'pcap-mode-set-tshark-single-packet-filter)
     (define-key kmap (kbd "r") 'pcap-mode-reload-file)
+    (define-key kmap (kbd "g") 'pcap-mode-clear-filter)
     (define-key kmap (kbd "\C-c \C-d") (lambda () (interactive)
-                                        (message "tshark filter \"%s\""
```
Comment (apconole) [inline pcap-mode.el:121]: Why these whitespace changes?  Please drop them.

File: pcap-mode.el [hunk 2]
```diff
@@ -362,6 +374,32 @@ Argument FILTER-VALUE corresponds to the exact set of filters passed to `pcap-mo
   (if pcap-mode-reload-pcap-when-filter-changes
       (pcap-mode-reload-file)))
 
+(defun pcap-mode-set-named-filter (filter-name)
```
Comment (apconole) [inline pcap-mode.el:390]: Please update the documentation string to include the purpose of FILTER-NAME

File: pcap-mode.el [hunk 3]
```diff
@@ -362,6 +374,32 @@ Argument FILTER-VALUE corresponds to the exact set of filters passed to `pcap-mo
...
+  "Choose a predefined filter and apply it.
+Filters are obtained from the `pcap-mode-dfilters-file'."
+  (interactive (list (ido-completing-read
```
Comment (apconole) [inline pcap-mode.el]: This requires ido mode, but there's no hint that such a dependency is required.  Please use completing-read.  You may also choose to write a function which detects the presence of ido-mode and prefers it vs. completing-read, with a fallback to completing-read.  Example below is completly untested:

(defun pcap-mode--completing-read ()
 "Returns the preferred completing read function."
  (if (fboundp ido-completing-read) ido-completing-read
    completing-read))

File: pcap-mode.el [hunk 4]
```diff
@@ -362,6 +374,32 @@ Argument FILTER-VALUE corresponds to the exact set of filters passed to `pcap-mo
...
+    (if (file-readable-p pcap-mode-dfilters-file)
+	(progn (with-temp-buffer
+		 (insert-file-contents pcap-mode-dfilters-file)
+		 (goto-char (point-min))
+		 (search-forward filter-name)
+		 (setq filter (buffer-substring-no-properties (+ 2 (point)) (line-end-position))))
+	       (pcap-mode-set-tshark-filter (concat "'" filter "'")))
+      (error "Can't read dfilters file: %s" pcap-mode-dfilters-file))))
+
+(defun pcap-mode-clear-filter nil
```
Comment (apconole) [inline pcap-mode.el:406]: Can you split this series into three commits.  Commit one can fix the group binding for the custom variables above, commit two can add the clear-filter, commit three can add the new scanning code.

ALSO, this function has no documentation.

File: pcap-mode.el [hunk 5]
```diff
@@ -362,6 +374,32 @@ Argument FILTER-VALUE corresponds to the exact set of filters passed to `pcap-mo
...
+  (interactive (list (ido-completing-read
+		      "Filter name: "
+		      (if (file-readable-p pcap-mode-dfilters-file)
+			  (with-temp-buffer
+			    (insert-file-contents pcap-mode-dfilters-file)
+			    (goto-char (point-min))
+			    (cl-loop while (re-search-forward "^\"\\([^\"]*\\)\"" nil t)
+				     collect (match-string 1)))
+			(error "Can't read dfilters file: %s" pcap-mode-dfilters-file)))))
+  (let (filter)
```
Comment (apconole) [inline pcap-mode.el:400]: This code looks eerily similar to the interactive completion code.  Can you abstract this out into something like pcap-mode--dfilters-selection or something?

File: pcap-mode.el [hunk 6]
```diff
@@ -112,14 +122,16 @@
     (define-key kmap (kbd "<return>") 'pcap-mode-view-pkt-contents)
     (define-key kmap (kbd "t") 'pcap-mode-toggle-tcp-conversation-view)
     (define-key kmap (kbd "f") 'pcap-mode-set-tshark-filter)
+    (define-key kmap (kbd "F") 'pcap-mode-set-named-filter)
     (define-key kmap (kbd "c") 'pcap-mode-search-frames)
     (define-key kmap (kbd "\C-u f")
       'pcap-mode-set-tshark-single-packet-filter)
     (define-key kmap (kbd "s") 'pcap-mode-set-tshark-single-packet-filter)
     (define-key kmap (kbd "r") 'pcap-mode-reload-file)
+    (define-key kmap (kbd "g") 'pcap-mode-clear-filter)
```
Comment (apconole) [inline pcap-mode.el:134]: I'm happy to see additional functions added to the default keymap.  Please update the README as well.

File: pcap-mode.el [hunk 7]
```diff
@@ -72,12 +72,22 @@
 ;;;###autoload
 (defcustom pcap-mode-tshark-filter ""
   "Filter to apply to tshark invocations."
-  :type 'string)
+  :type 'string
```
Comment (apconole) [inline pcap-mode.el:77]: Please put this change, and the similar one below (for pcap-mode-tshark-single-packet-filter) in a separate commit.

**Review by apconole — CHANGES_REQUESTED** [review]
Thanks for this.  I think after a round of fixups this will be acceptable.

Please also make sure to add a signed-off-by line, to acknowledge that your contribution is covered under the GNU GPL (git commit -s), and update the changelog in the pcap-mode.el file.

File: pcap-mode.el (hunk 7 above)
Comment (vapniks) [reply pcap-mode.el:77]: done

Comment (vapniks) [reply pcap-mode.el:77]: Have updated the changelog, but forgot to add signed off lines, sorry (aren't my changes automatically covered by the GPL anyway?).

File: pcap-mode.el (hunk 6 above)
Comment (vapniks) [reply pcap-mode.el:134]: done

File: pcap-mode.el (hunk 1 above)
Comment (vapniks) [reply pcap-mode.el:121]: This is an artifact of `aggressive-indent` mode, some of your whitespace was non-standard (opening parentheses should always be to the right of opening parentheses of parent sexp's).

File: pcap-mode.el (hunk 2 above)
Comment (vapniks) [reply pcap-mode.el:390]: done

File: pcap-mode.el (hunk 3 above)
Comment (vapniks) [reply pcap-mode.el]: done (differently)

File: pcap-mode.el (hunk 5 above)
Comment (vapniks) [reply pcap-mode.el:400]: done, (but not sure it was worth it).

File: pcap-mode.el (hunk 4 above)
Comment (vapniks) [reply pcap-mode.el:406]: done

**Comment by apconole** [comment]
Thanks for your contribution!  I manually cherry picked them to avoid the big 'add then revert' commit in the middle.

**Pull request closed without merging**